resolve issue 5652 to support multiple reference genomes

khzhu opened the pull request:
## What? Why?
Resolve issue [https://github.com/cBioPortal/cbioportal/issues/5652](url) to support multiple reference genomes
- [x] remove **_cytoband_** , **chromosome**, and **_length_** from **_gene_** entity
- [x] fetch reference genome genes of studies of interest using APIs
- [x] use **_cytoband_** from **_reference genome genes_** rather than from **_genes_**
- [x] load chromosome sizes of different reference genomes from JSON file for segments/mutations plots on the patient view
- [x] use **_chr_** from mutation entity instead of calculating from **_cytoband_**
- [x] disable **_View Summary_** when studies selected come from different reference genomes

Fix #

Changes proposed in this pull request:
- a
- b

# Checks
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Airbnb React/JSX Style guide](https://github.com/airbnb/javascript/tree/master/react).
- [ ] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [GifGrabber](http://www.gifgrabber.com/).

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). If you are part of the
cBioPortal organization, notify the approprate team (remove inappropriate):

@cBioPortal/frontend

If you are not part of the cBioPortal organization look at who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them here:
#

khzhu commented:
Hi @adamabeshouse @onursumer , can I please have your review on the PR? I mades some changes in the codebase you have been working on. It will be great to have some feedback. Please let me if things are unclear or you need more information. Thanks!

adamabeshouse commented:
@khzhu I'm not familiar with this, I'll leave it to @onursumer

khzhu commented:
@adamabeshouse Okay. Would you be able to point one, since I need two reviewers. Thanks!

khzhu commented:
@adamabeshouse , just found I only need one approving review. Great!

khzhu commented:
@alisman @onursumer , the code is ready for your review. Thanks!

alisman commented on src/pages/patientView/copyNumberAlterations/CopyNumberTableWrapper.tsx:
```diff
@@ -57,6 +58,12 @@ export default class CopyNumberTableWrapper extends React.Component<ICopyNumberT
         enableCivic: false
     };
 
+    getCytoband(hugoGeneSymbol:string) {
```
i would call this getCytobandForGene

khzhu replied:
will change.

khzhu replied:
changed to getCytobandForGene

alisman commented on src/pages/patientView/copyNumberAlterations/CopyNumberTableWrapper.tsx:
```diff
@@ -57,6 +58,12 @@ export default class CopyNumberTableWrapper extends React.Component<ICopyNumberT
+    getCytoband(hugoGeneSymbol:string) {
+        const refGene =  _.find(this.props.refGenes,(gene:ReferenceGenomeGene)=>
+            gene.hugoGeneSymbol===hugoGeneSymbol);
+        return refGene? refGene.cytoband: "-";
```
returing "-" seems like a presentation choice.  I would return null or undefined and then have view code specify what to show in that case.

khzhu replied:
agree.

khzhu replied:
replaced "-" with undefined.

alisman commented on src/pages/patientView/genomicOverview/Tracks.tsx:
```diff
@@ -31,7 +31,8 @@ export default class Tracks extends React.Component<TracksPropTypes, {}> {
         // --- chromosome chart ---
-        var chmInfo = tracksHelper.getChmInfo();
+        const chmInfo = tracksHelper.getChmInfo( this.props.mutations?
+            this.props.mutations[0]!.ncbiBuild: "GRCh37");
```
if that is a default chromosome, you should make it a constant.  e.g. DEFAULT_CHROMOSOME at top of file

khzhu replied:
agree. Will make **GRCh37** DEFAULT_GENOME_BUILD for now.

khzhu replied:
made GRCh37 a const.

onursumer commented on src/pages/resultsView/cnSegments/CNSegments.tsx:
```diff
@@ -51,9 +51,15 @@ export default class CNSegments extends React.Component<{ store: ResultsViewPage
+    @computed get genome() {
+        const study = this.props.store.studies.result?
+            this.props.store.studies.result[0]: undefined;
+        return study? study.referenceGenome: "hg19";
```
maybe create a constant like `DEFAULT_GENOME`

khzhu replied:
good idea.

khzhu replied:
defined DEFAULT_GENOME in IGVUtils file.

alisman commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +56,38 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
-    sel.hg19 = [0,249250621,243199373,198022430,191154276,180915260,171115067,159138663,146364022,141213431,135534747,135006516,133851895,115169878,107349540,102531392,90354753,81195210,78077248,59128983,63025520,48129895,51304566,155270560,59373566];
-    sel.total = 3095677412;
-    sel.perc = getChmEndsPerc(sel.hg19,sel.total);
+    sel.genomeRef = {};
```
you could put this property inside object above

khzhu replied:
good point. thanks!

onursumer commented on src/shared/components/lollipopMutationPlot/LollipopMutationPlot.tsx:
```diff
@@ -286,7 +286,7 @@ export default class LollipopMutationPlot extends React.Component<ILollipopMutat
             this.props.store.transcriptsByTranscriptId[this.props.store.activeTranscript] &&
             this.props.store.transcriptsByTranscriptId[this.props.store.activeTranscript].proteinLength) ||
-            Math.round(this.props.store.gene.length / 3);
+            0;
```
can we set it to some non-zero value like 10?

khzhu replied:
okay, use 10.

khzhu replied:
replaced 0 with 10.

alisman commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +56,38 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+
+    }
+    const chromSizes = require('./chromSizes.json');
+    for (let i = 0; i < chromSizes.length; i++) {
```
instead of using loop break, use _.find to get first matching.  also, what happens if it doesn't find anything?

khzhu replied:
good point! Will use _.find. Add if nothing found then default hg19 size to be returned.

khzhu replied:
change made. replaced old style javascript code with find and reduce.

alisman commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +56,38 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+            break;
+        }
+    }
+    sel.total = 0
```
put this up in the definition of sel at top of this routine

khzhu replied:
will do.

khzhu replied:
done.

alisman commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +56,38 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+        }
+    }
+    sel.total = 0
+    for (let i=0; i < sel.genomeRef.length; i++ ) {
```
use lodash _.sum for this

khzhu replied:
cool! will use _.sum. thanks!

khzhu replied:
used .reduce method.

khzhu commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -643,6 +669,23 @@ export class PatientViewPageStore {
     }, undefined);
 
+
+    readonly refGenes = remoteData<ReferenceGenomeGene[]>({
```
sure, will change.

alisman replied:
@jj @khzhu  I think our naming scheme needs to carefully distinguish this from the genes collection.  If I understand correctly, there can be multiple ReferenceGenomeGene's per Gene.  Do they represent a particular position on a particular reference genome?  Maybe the name could reflect that, like, ReferenceGenomeGeneDefinition  or ReferenceGenomeGeneVariant or ReferenceGenomeGenePosition?  Something that signals the difference with Gene concept.

khzhu replied:
@alisman , you are right, a gene can have multiple reference genome genes. The reference genome gene are still genes but with different cytogenetic locations. I do not know if we need to distinguish genes from reference genes or just pick up right coordinates/cytobands when plot variants.  @JJ what do you think?

jjgao replied:
@khzhu Is it true that you are using `ReferenceGenomeGene` to reflect the database name / api variable? If so, maybe we should keep it.

khzhu replied:
@JJ, yes, that is right. I see Aaron's points. ReferenceGenomeGene object essentially encapsulates a set of properties of a gene from multiple genomes. ReferenceGenomeGeneDefinition or ReferenceGenomeGeneProperties might less confuse people.

khzhu commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2079,6 +2081,11 @@ export class ResultsViewPageStore {
         return parsedQString.query.session_id;
     }
 
+    private getCytoBand(hugoGeneSymbol:string):string {
```
will use getCytoband to be consistent.

khzhu replied:
used getCytoband.

khzhu commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2156,6 +2163,28 @@ export class ResultsViewPageStore {
         }
     });
 
+    readonly refGenes = remoteData<ReferenceGenomeGene[]>({
```
agree.

khzhu replied:
used referenceGenes.

khzhu commented on src/pages/patientView/genomicOverview/chromSizes.json:
```diff
@@ -0,0 +1,14 @@
+[
```
will change the name.

khzhu replied:
renamed chromosome size file.

alisman commented on src/shared/components/query/CancerStudySelector.tsx:
```diff
@@ -134,6 +134,20 @@ export default class CancerStudySelector extends React.Component<ICancerStudySel
     });
 
+    @computed
+    get checkReferenceGenome() {
```
if i understand correctly, this checks if all selected studies have same referencegenome.  It true, lets call it checkForSingleReferenceGenome

khzhu replied:
yes, you are right. Will change.

khzhu replied:
renamed to checkForSingleReferenceGenome

alisman commented on src/shared/components/query/CancerStudySelector.tsx:
```diff
@@ -134,6 +134,20 @@ export default class CancerStudySelector extends React.Component<ICancerStudySel
+    get checkReferenceGenome() {
+        let genomeValid = true;
+        if (this.store.selectableSelectedStudies.length > 0) {
+            const referenceGenome = this.store.selectableSelectedStudies[0].referenceGenome;
```
i think you could map to an array of all referneceGenomes, then call _.uniq on that array, then measure length > 1 to determine validity of query

alisman replied:
also, do we allow user to submit query when checkReferenceGenome is false (mixed genomes) ? Or do we block them from doing so, as we do if they select too many studies?

khzhu replied:
we do not allow merge studies from different genomes.

alisman commented on src/shared/components/query/CancerStudySelector.tsx:
```diff
@@ -225,9 +239,12 @@ export default class CancerStudySelector extends React.Component<ICancerStudySel
                                 {() => {
 
-                                    const studyLimitReached = (this.store.selectableSelectedStudyIds.length > 50);
+                                    const studyLimitReached = (!this.checkReferenceGenome) ||
```
seems odd to combine these two invalid "states" no?   I think we should have a separate message for studies from different reference genome

khzhu replied:
agree. Let's have  a new message.

khzhu replied:
added a new message for multiple reference genomes.

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -707,8 +709,8 @@ export class ResultsViewPageStore {
-            if (this.genes.result) {
-                return Promise.resolve(_.reduce(_.uniq(this.genes.result.map(g => g.chromosome)), (map: { [chromosome: string]: MobxPromise<CopyNumberSeg[]> }, chromosome: string) => {
+            if (this.refGenes.result) {
+                return Promise.resolve(_.reduce(_.uniq(this.refGenes.result.map(g => g.chromosome)), (map: { [chromosome: string]: MobxPromise<CopyNumberSeg[]> }, chromosome: string) => {
```
this code is kind of hard to read.  i would first make a variable uniqueChromosommes and assign map to it.  Then call reduce on variable.

khzhu replied:
good idea! will make changes.

khzhu replied:
made a variable **_uniqueReferenceGenes_** and call reduce on the variable.

alisman commented on src/shared/lib/StoreUtils.ts:
```diff
@@ -85,6 +99,31 @@ export async function fetchGenes(hugoGeneSymbols?: string[],
+{
+    if (hugoGeneSymbols && hugoGeneSymbols.length) {
+        const order = stringListToIndexSet(hugoGeneSymbols);
+        return _.sortBy(await client.fetchGenesUsingPOST_1({
```
@onursumer noticed this but it seem slike fetchGenesUsingPost is overloaded (on backend) which produces these indexed methods on client.   We need to better distinguish referenceGenomeGenes from genes.

onursumer replied:
@alisman @khzhu I have added the relevant review for this one here: https://github.com/cBioPortal/cbioportal/pull/5891

khzhu replied:
thanks, @onursumer! Those notes I put on schema diagram might help with the review.

onursumer commented on src/pages/resultsView/cnSegments/CNSegments.tsx:
```diff
@@ -7,7 +7,7 @@ import {Nav, NavItem} from "react-bootstrap";
 
```
We have another `CNSegments` tab for study view: `src/pages/studyView/tabs/CNSegments.tsx`. We may want to update that one as well with the genome information.

khzhu replied:
**_CNSegments.tsx_** page is not affected, since **_IntegrativeGenomicsViewer_** is the one load genome reference fa file from s3.

inodb replied:
should IGV be given some parameter to pick the correct reference genome?

khzhu commented:
Hi @alisman @onursumer , I've addressed all comments from your reviews.Please take a final review on this PR and let me know if anything else need to be done. Thanks again for your time!

alisman commented on src/pages/patientView/clinicalInformation/PatientViewPageStore.ts:
```diff
@@ -1,805 +1,822 @@
+        onError: (err: Error) => {
+            // fail silently
+        }
+
```
looks like you have reformatted this file, causing every line to be updated?

khzhu replied:
the file had carriage return in windows format ^M, got red ! warning in WebStorm, so changed to linux CRLF. Should I roll back the change?

alisman replied:
were you able to roll this back?   I just checked PatientVIewPageStore in master and webstorm says it's set to LF, like all rest of files, so perhaps try rebasing?

khzhu replied:
will do a rebase. thanks!

alisman commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +56,36 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+    sel.genomeRef = {};
+    sel.total = 0
+
+    const chromSizes:chromosomeSizeType[] = require('./chromosomeSizes.json');
```
you should put this require at top of file

alisman commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +56,36 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+    const genomeSize = chromSizes.find(c => c.build === referenceGenome);
+    if (genomeSize) {
+        sel.genomeRef = genomeSize.size;
+        sel.total = sel.genomeRef.reduce((a: number, b: number) => a + b);
```
use _.sum

khzhu replied:
changed.

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -721,14 +723,14 @@ export class ResultsViewPageStore {
-            if (this.genes.result) {
-                return Promise.resolve(_.reduce(_.uniq(this.genes.result.map(g => g.chromosome)), (map: { [chromosome: string]: MobxPromise<CopyNumberSeg[]> }, chromosome: string) => {
+            if (this.referenceGenes.result) {
+                const uniqueReferenceGenes = _.uniq(this.referenceGenes.result.map(g => g.chromosome));
```
is it right that we are assigning "chromosome"s to a variable called referenceGenes?

khzhu replied:
we are retrieving copy number segments by chromosomes in reference genome genes.

adamabeshouse replied:
aaron is right, uniqueReferenceGenes contains chromosomes. it should be called uniqueReferenceGeneChromosomes or something like that

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -721,14 +723,14 @@ export class ResultsViewPageStore {
-                return Promise.resolve(_.reduce(_.uniq(this.genes.result.map(g => g.chromosome)), (map: { [chromosome: string]: MobxPromise<CopyNumberSeg[]> }, chromosome: string) => {
+            if (this.referenceGenes.result) {
+                const uniqueReferenceGenes = _.uniq(this.referenceGenes.result.map(g => g.chromosome));
+                return Promise.resolve(uniqueReferenceGenes.reduce(
```
lets assign result of reduce to local variable first and then resolve it.  hard to read

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2095,6 +2097,15 @@ export class ResultsViewPageStore {
+        const referenceGenes = this.referenceGenes!.result;
+        if (referenceGenes) {
+            const gene = _.find(referenceGenes, (g: ReferenceGenomeGene) => g.hugoGeneSymbol === hugoGeneSymbol);
+            return gene ? gene.cytoband : "";
```
should we really use empty string here?  seems it would be better to use null until actual use

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2172,6 +2183,28 @@ export class ResultsViewPageStore {
+            this.genes
+        ],
+        invoke: async () => {
+            const queryGenes = this.genes.result!.map((g:Gene)=>g.hugoGeneSymbol.toUpperCase());
```
is there a reason why we have to uppercase them?  seems like not a good practice.

adamabeshouse replied:
we do this everywhere. it is unpleasant but... it is

alisman commented on src/pages/groupComparison/GroupComparisonStore.ts:
```diff
@@ -397,6 +399,26 @@ export default class GroupComparisonStore {
+            return fetchReferenceGenomeGenes(this.studies.result[0].referenceGenome,
+                mutGenes.concat(cnvGenes));
+        },
+        onError:(err)=>{
```
we should actually get rid of onError unless the error is specifically dealt with in your code.  Sorry, we are updating our practice around errors.

alisman replied:
getting rid of onError will allow global error handling to detect and show error message

alisman commented on src/pages/resultsView/enrichments/AlterationEnrichmentsContainer.tsx:
```diff
@@ -70,7 +73,19 @@ export default class AlterationEnrichmentContainer extends React.Component<IAlte
+            if (refGenes && refGenes.length > 0) {
+                const gene = refGenes.find(g => g.entrezGeneId === a.entrezGeneId);
+                if (gene) {
+                    a.cytoband = gene.cytoband;
```
@adamabeshouse wanted to get your opinion here.  Seems like we are extending api data here.  One question I have is why AlterationEnrichment type has a cytoband property if it is not populated?  or are overriding it here?  It concerns me that we mutating global data source based on referenceGenes property of this component.  If this is accetable to do a general level (application wide) then we should do this in the store to ensure that all consumers of the data see the same thing under all circumstances.

adamabeshouse replied:
agreed. we should do it at the top.

alisman commented on src/pages/studyView/tabs/CNSegments.tsx:
```diff
@@ -69,6 +69,12 @@ export default class CNSegments extends React.Component<{ store: StudyViewPageSt
+    @computed get genome() {
+        const study = this.props.store.allPhysicalStudies.result?
+            this.props.store.allPhysicalStudies.result[0]: undefined;
+        return study? study.referenceGenome: DEFAULT_GENOME;
```
i think there should be a space before ? and another before the colon in this ternary.

alisman replied:
if the reference genome is "global," meaning there is nothing specific to CNSegments, then it probably belongs in the store rather than here?

adamabeshouse replied:
agreed, `genome()` should be put in resultsviewpagestore

alisman commented on src/shared/components/query/CancerStudySelector.tsx:
```diff
@@ -134,6 +134,12 @@ export default class CancerStudySelector extends React.Component<ICancerStudySel
         );
     });
 
+    @computed
```
seems like this belongs in the store

adamabeshouse replied:
agreed. also please rename to `onlyOneReferenceGenome` or something like that, to agree with the getter syntax.

alisman commented on src/shared/components/query/CancerStudySelector.tsx:
```diff
@@ -225,21 +231,25 @@ export default class CancerStudySelector extends React.Component<ICancerStudySel
+                                    const studyLimitReached = this.store.selectableSelectedStudyIds.length > 50;
+                                    const multipleReferenceGenomes = (!this.checkForSingleReferenceGenome);
+                                    let tooltipMessage = <span>Open summary of selected studies in a new window.</span>;
+                                    if (studyLimitReached) {
```
i think these two ifs should be if/else and probably studyLimitReached should show first.

adamabeshouse replied:
agreed

alisman commented on src/shared/components/query/CancerStudySelector.tsx:
```diff
@@ -225,21 +231,25 @@ export default class CancerStudySelector extends React.Component<ICancerStudySel
                                             >
 
-                                                <Button bsSize="xs" disabled={studyLimitReached} bsStyle="primary"
+                                                <Button bsSize="xs" disabled={studyLimitReached||multipleReferenceGenomes} bsStyle="primary"
```
lets take this condition out and assign it to a local variable of the method just to make it super clear.

alisman commented on src/shared/components/query/CancerStudySelector.tsx:
```diff
@@ -225,21 +231,25 @@ export default class CancerStudySelector extends React.Component<ICancerStudySel
                             {!!(!this.store.cancerTypes.isPending && !this.store.cancerStudies.isPending) && (
                                 <Observer>
                                     {() => {
+                                    const studyLimitReached = this.store.selectableSelectedStudyIds.length > 50;
```
@khzhu i realized that this button has been eliminated (replaced by "forked mode" button).  You should rebase and then reapply this logic to the forked button.  Let me know if you need help. SHould be pretty clear when you rebase (or if you look at latest public portal)

adamabeshouse commented on src/shared/lib/IGVUtils.ts:
```diff
@@ -3,6 +3,7 @@ import {CopyNumberSeg} from "shared/api/generated/CBioPortalAPI";
 import {TrackProps} from "shared/components/igv/IntegrativeGenomicsViewer";
 
 export const WHOLE_GENOME = "all";
+export const DEFAULT_GENOME = "hg19";
```
if the reference genome is a page wide setting, I don't think this should go in IGVUtils. It should go in ResultsViewPageStoreUtils

adamabeshouse commented on src/pages/patientView/copyNumberAlterations/CopyNumberTableWrapper.tsx:
```diff
@@ -57,6 +58,12 @@ export default class CopyNumberTableWrapper extends React.Component<ICopyNumberT
     };
 
+    getCytobandForGene(hugoGeneSymbol:string) {
+        const refGene =  _.find(this.props.referenceGenes,(gene:ReferenceGenomeGene)=>
```
this is a linear search through props.referenceGenes every time you want to call this method, which is inefficient and could become a bit slow for a query with a lot of rows and a lot of genes.

Instead,

```
@computed get hugoGeneSymbolToCytoband() {
    if (this.props.referenceGenes) {
        return this.props.referenceGenes.reduce((map, next)=>{
            map[next.hugoGeneSymbol] = next.cytoband;
        });
    } else {
        return {};
    }
}
```

adamabeshouse commented on src/pages/patientView/genomicOverview/Tracks.tsx:
```diff
@@ -14,6 +14,8 @@ interface TracksPropTypes {
     width:number;
 }
 
+export const DEFAULT_CHROMOSOME="GRCh37";
```
just want to confirm this belongs in `Tracks`, so it is a `Tracks`-specific constant that's not used elsewhere. otherwise, it should go in a utils file that's more generally used.

adamabeshouse commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +58,33 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+    [ "mm10", "GRCm38" ]
+]);
+
+export type chromosomeSizeType = {
```
generally it's not good to name a type using the word "type". also, types should be capitalized. would `ChromosomeSizes` work?

adamabeshouse commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +58,33 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+    size: number[]
+};
+
+export function getChmInfo(genomeBuild:string) {
```
as in a comment above, it would be more efficient just to build a map genomeBuild->chromosome info, then index into that map, rather than calling a function.

adamabeshouse commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +58,33 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+    if (!referenceGenome || referenceGenome === "") {
+        referenceGenome = DEFAULT_CHROMOSOME;
+    }
+    const genomeSize = chromSizes.find(c => c.build === referenceGenome);
```
better to build a map: genomeBuild->genomeSize. then index into that. otherwise, every time you make this function call, you're doing a linear time scan of chromosomeSizes.json.

adamabeshouse commented on src/pages/resultsView/ResultsViewPage.tsx:
```diff
@@ -305,7 +305,7 @@ export default class ResultsViewPage extends React.Component<IResultsViewPagePro
                             (<Network genes={store.genes.result!}
                                       profileIds={store.rvQuery.selectedMolecularProfileIds}
-                                      cancerStudyId={store.studies.result[0].studyId}
+                                      cancerStudyId={(store.studies.result.length > 0) ?store.studies.result![0].studyId:""}
```
here and everywhere else, ternary spaces should be consistent. just looks more readable.
either `a?b:c` or `a ? b : c`, I think the second one is better. but not `a ?b:c`

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -795,14 +797,14 @@ export class ResultsViewPageStore {
+                return Promise.resolve(uniqueReferenceGenes.reduce(
+                    (map: { [chromosome: string]: MobxPromise<CopyNumberSeg[]> }, chromosome: string) => {
+                        map[chromosome] = remoteData<CopyNumberSeg[]> ({
+                            invoke: () => fetchCopyNumberSegmentsForSamples(this.samples.result, chromosome)});
```
please put the` });` on the next line, it looks cleaner

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2190,6 +2191,15 @@ export class ResultsViewPageStore {
         return parsedQString.query.session_id;
     }
 
+    private getCytoband(hugoGeneSymbol:string):string {
```
should be `readonly referenceGeneHugoSymbolToCytoband =  remoteData`. need to avoid these linear time scans in function calls when they can be precomputed once into a map and then constant time lookup every time afterwards.

adamabeshouse commented on src/pages/resultsView/cnSegments/CNSegments.tsx:
```diff
@@ -51,9 +51,15 @@ export default class CNSegments extends React.Component<{ store: ResultsViewPage
 
+    @computed get genome() {
+        const study = this.props.store.studies.result?
+            this.props.store.studies.result[0]: undefined;
```
pull out and put in the store, as discussed below in CNSegmentsTab

khzhu commented:
I have done my corrections. All comments have been addressed. For the very few ones not corrected I explained why I decided to keep them the way they are. 
Please let me know anything else I need to do to get your approval. Thanks again for those informative and detailed suggestions!

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2267,6 +2270,50 @@ export class ResultsViewPageStore {
+        }
+    });
+
+    @computed get hugoGeneSymbolToCytoband() {
```
(1) After discussion over time, @alisman and I have determined that in cases like this (taking this as an example): it should be a remoteData which awaits this.referenceGenes, rather than a computed which checks the referenceGenes status.
(2) Isn't the name unclear as written, because it does not distinguish from the cytoband of a gene not on the selected reference genome? I think it should have the word "reference gene" in there somewhere
(3) `hugoGeneSymbolToCytoband` and `hugoGeneSymbolToChromosome` are redundant - should just be `hugoGeneSymbolToReferenceGene`, then users can do `hugo...Gene[gene].cytoband` or `.chromosome`

`referenceGeneHugoGeneSymbolToCytoband` is better, to distinguish from the query genes that are not from the selected reference genome. Also, please make this a `remoteData` which awaits `this.referenceGenes`. Please also make `hugoGeneSymbolToChromosome`

khzhu commented:
@alisman , I've pushed all changes (done so far) back to the repo for you to take a look. Please let me know if anything else you need from me. Thanks again for the help!

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -795,14 +798,15 @@ export class ResultsViewPageStore {
-
-                    return map;
-                }, {}));
+            if (this.referenceGenes.result) {
```
you need to await referenceGenes if you use it

kalletlak commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2267,6 +2270,93 @@ export class ResultsViewPageStore {
+            this.studies,
+            this._mutationEnrichmentDataWithoutCytoband,
+            this._copyNumberHomdelEnrichmentData,
+            this._copyNumberAmpEnrichmentData,
```
this would be fetching both mutation and cna enrichments data together. this might cause delay in loading the page that is viewed. @adamabeshouse any suggestions?

adamabeshouse replied:
good point. one way to solve it is fetch and deal with mutation enrichment reference genes separately from cna enrichment reference genes.

khzhu commented:
new comments from @adamabeshouse @kalletlak have been addressed. Please let me know anything else I might have missed. Thank you all again for those cool tips/suggestions!

adamabeshouse commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +58,40 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+
+const chromSizes:ChromosomeSizes[] = chromosomeSizes;
+
+const referenceGenomeSizes:{[genomeBuild:string]:number[]} =
```
(1) whats the point of having chromSizes when all it does is assign a reference to chromosomeSizes?
(2) this is easier to read as _.keyBy(chromosomeSizes, s=>s.genomeBuild)

adamabeshouse commented on src/pages/patientView/genomicOverview/tracksHelper.ts:
```diff
@@ -56,11 +58,40 @@ function getChmEndsPerc(chms: Array<any>, total: any) {
+    chromosomeSize: number[]
+};
+
+const chromSizes:ChromosomeSizes[] = chromosomeSizes;
```
(1) whats the point of having chromSizes when all it does is assign a reference to chromosomeSizes?

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2267,6 +2271,81 @@ export class ResultsViewPageStore {
+        ],
+        invoke: ()=>{
+            // build reference gene map
+            const result:{[hugosymbol:string]:ReferenceGenomeGene} =
```
theres no need to save `result`, just return `Promise.resolve(_.keyBy(...))`

if there's typing issues, you can annotate the type above:
`remoteData<{[hugoSymbol:string]:ReferenceGenomeGene}>({ ...`

adamabeshouse commented on src/pages/resultsView/cnSegments/CNSegments.tsx:
```diff
@@ -52,10 +52,9 @@ export default class CNSegments extends React.Component<{ store: ResultsViewPage
-            this.props.store.genes.result.find(g => g.hugoGeneSymbol === this.activeLocus) : undefined;
-
-        return gene ? gene.chromosome : undefined;
+        if (this.props.store.hugoGeneSymbolToReferenceGene.result) {
```
generally in this case we would want to use a remoteData - await hugoGeneSymbolToReferenceGene and then use it. unless it genuinely doesnt matter and `undefined` is an equally valid value. In the latter case, you would still need to `return undefined` explicitly

adamabeshouse commented on src/pages/patientView/copyNumberAlterations/CopyNumberTableWrapper.tsx:
```diff
@@ -48,15 +50,27 @@ type ICopyNumberTableWrapperProps = {
 
+    @computed get hugoGeneSymbolToCytoband() {
+        // build reference gene map
+        const result:{[hugosymbol:string]:string} =
```
you should do a check for referenceGenes to be defined here, otherwise could end up with a bug. Unless referenceGenes is mandatory, in which case remove the `?` from the prop.

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -792,17 +795,19 @@ export class ResultsViewPageStore {
-
-                    return map;
-                }, {}));
+            if (this.referenceGenes.result) {
```
if you're awaiting referenceGenes, then `result` will always be there. no need to check

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2891,7 +2972,7 @@ export class ResultsViewPageStore {
     });
 
-    readonly copyNumberHomdelEnrichmentData = makeEnrichmentDataPromise({
+    readonly _copyNumberHomdelEnrichmentData = makeEnrichmentDataPromise({
```
WithoutCytoband, for clarity and coherence

adamabeshouse replied:
nvm: see my comment here https://github.com/cBioPortal/cbioportal-frontend/pull/2180/files#r310723393

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2902,7 +2983,7 @@ export class ResultsViewPageStore {
     );
 
-    readonly copyNumberAmpEnrichmentData = makeEnrichmentDataPromise({
+    readonly _copyNumberAmpEnrichmentData = makeEnrichmentDataPromise({
```
WithoutCytoband

adamabeshouse replied:
nvm: see my comment here https://github.com/cBioPortal/cbioportal-frontend/pull/2180/files#r310723393

adamabeshouse commented on src/pages/groupComparison/GroupComparisonStore.ts:
```diff
@@ -397,6 +399,22 @@ export default class GroupComparisonStore {
+            const cnvGenes = this.copyNumberData.result!.map(
+                (a:AlterationEnrichment)=>a.hugoGeneSymbol.toUpperCase());
+            return fetchReferenceGenomeGenes(this.studies.result[0].referenceGenome,
+                mutGenes.concat(cnvGenes));
```
seems like you have to do the same thing here, right? or else the cytoband column will be empty in group comparison

adamabeshouse replied:
nvm: see my comment https://github.com/cBioPortal/cbioportal-frontend/pull/2180/files#r310723393

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -2267,6 +2271,81 @@ export class ResultsViewPageStore {
+
+    readonly mutationEnrichmentData  = remoteData({
+        await: ()=>[
+            this._mutationEnrichmentDataWithoutCytoband,
```
since you're now just doing `fetchAllReferenceGenomeGenes`, theres no need to do this thing anymore where we calculate the cytoband afterwards. see what I mean? so you can just put it back to the way you had it before, where you add the cytoband inside makeEnrichmentsDataPromise
